File: MVP.Sample/MVP.Base/BasePresenter/BasePresenter.cs
using MVP.Base.BaseRepository;
using MVP.Base.BaseView;
using MVP.Base.Common;

namespace MVP.Base.BasePresenter
{
    /// <summary>
    /// 
    /// </summary>
    public abstract class BasePresenter<TView, TRepository> : IBasePresenter
        where TView : class, IBaseView
        where TRepository : IBaseRepository
    {
        #region IView
        protected readonly TView View;
        #endregion

        #region Repository

        protected TRepository Repository;
        #endregion

        #region IBasePresenter

        /// <summary>
        /// Initializes a new instance of the <see cref="BasePresenter&lt;TView, TRepository&gt;"/> class.
        /// </summary>
        /// <param name="view">The view.</param>
        protected BasePresenter(TView view)
        {
            View = view;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="BasePresenter{TView,TRepository}"/> class.
        /// </summary>
        /// <param name="view">The view.</param>
        /// <param name="repository">The repository.</param>
        protected BasePresenter(TView view, TRepository repository)
        {
            View = view;
            Repository = repository;
        }

        #endregion

        #region Actions

        protected abstract void ViewAction();
        protected abstract void UpdateAction();
        protected abstract void DeleteAction();
        protected abstract void InsertAction();
        protected abstract void InitializeAction();
        protected abstract void SearchAction();

        /// <summary>
        /// Does the action.
        /// </summary>
        public virtual void DoAction()
        {
            switch (View.PageMode)
            {
                case PageMode.None:
                    InitializeAction();
                    break;
                case PageMode.Insert:
                    InsertAction();
                    break;
                case PageMode.Delete:
                    DeleteAction();
                    break;
                case PageMode.Update:
                    UpdateAction();
                    break;
                case PageMode.View:
                    ViewAction();
                    break;
                case PageMode.Search:
                    SearchAction();
                    break;
            }
        }

        /// <summary>
        /// Validates this instance.
        /// </summary>
        public abstract void Validate();

        #endregion
    }
}

File: trunk/DotNetNuke/DNN.MVP.Base/BasePresenter/BasePresenter.vb
Imports DNN.MVP.Base.Common
Imports DNN.MVP.Base.BaseView

Namespace BasePresenter
    Public MustInherit Class BasePresenter(Of TView As IBaseView, TRepository As Class)
        Implements IBasePresenter

#Region "View"
        Protected View As TView
#End Region

#Region "Repository"
        Protected Overridable ReadOnly Property Repository() As TRepository
            Get
                Return Nothing
            End Get
        End Property
#End Region

#Region "Constructors"
        Protected Sub New(ByVal tView As TView)
            View = tView
        End Sub
#End Region

#Region "Actions"
        Public MustOverride Sub ViewAction()
        Public MustOverride Sub UpdateAction()
        Public MustOverride Sub DeleteAction()
        Public MustOverride Sub InsertAction()
        Public MustOverride Sub InitializeAction()
        Public MustOverride Sub SearchAction()

        Public Overridable Sub DoAction() Implements IBasePresenter.DoAction
            Select Case View.PageMode
                Case PageMode.NONE
                    InitializeAction()
                    Exit Select
                Case PageMode.VIEW
                    ViewAction()
                    Exit Select
                Case PageMode.INSERT
                    InsertAction()
                    Exit Select
                Case PageMode.UPDATE
                    UpdateAction()
                    Exit Select
                Case PageMode.DELETE
                    DeleteAction()
                    Exit Select
                Case PageMode.SEARCH
                    SearchAction()
                    Exit Select
            End Select
        End Sub
#End Region

#Region "Actions"
        Public MustOverride Sub Validate() Implements IBasePresenter.Validate
#End Region

    End Class
End Namespace

